### PR TITLE
[SMTChecker] Small refactoring of assignment

### PR DESCRIPTION
**Pull request by blishko**
This PR suggests some more refactoring of the code handling assignments in SMTChecker.
It centralizes the handling of defining given SMT Expression as a new value for the given Solidity Expression in the method `SMTEncoder::assign(Expression const&, smtutil::Expression const&, Type const*)` with the aim of reducing code duplication.

### CODE REVIEW

File: libsolidity/formal/SMTEncoder.cpp [hunk 1]
```diff
@@ -1974,8 +1932,29 @@ void SMTEncoder::assignment(
...
 	)
 		indexOrMemberAssignment(*left, _right);
-	else if (isEmptyPush(*left))
-		arrayPushPopAssign(*left, _right);
+	else if (auto const* funCall = dynamic_cast<FunctionCall const*>(left))
+	{
+		if (
+			auto funType = dynamic_cast<FunctionType const*>(funCall->expression().annotation().type);
+			funType && funType->kind() == FunctionType::Kind::ArrayPush
+		) {
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]: `{` should be on the next line

**Comment by leonardoalt** [comment]
@blishko does this PR depend on the query order PR?

**Comment by blishko** [comment]
Not exactly, but it actually changes the encoding slightly (it removes one redundant re-definition of an expression, I believe).
So it requires update to the expected output of the command line tests.

I thought it would be better to actually properly see the changes in the expected output, and *that* depends on the query order PR.

File: libsolidity/formal/SMTEncoder.cpp [hunk 2]
```diff
@@ -1596,54 +1568,7 @@ void SMTEncoder::arrayPop(FunctionCall const& _funCall)
...
-
-	if (auto const* id = dynamic_cast<Identifier const*>(expr))
-	{
-		auto varDecl = identifierToVariable(*id);
-		solAssert(varDecl, "");
-		if (varDecl->hasReferenceOrMappingType())
-			resetReferences(*varDecl);
-		m_context.addAssertion(m_context.newValue(*varDecl) == _array);
-		m_context.expression(*id)->increaseIndex();
-		defineExpr(*id,currentValue(*varDecl));
```
Comment (blishko) [inline libsolidity/formal/SMTEncoder.cpp:1622]: The reason for the encoding change is here, I believe: lines 1612-1614 are replaced by simple `assignment` call for a `VariableDeclaration` (see line 1928 of the new version (1998 of the old version)).

File: libsolidity/formal/SMTEncoder.cpp (hunk 1 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp]: Fixed.

**Comment by leonardoalt** [comment]
@blishko I think we should merge this. Can you rebase?

**Comment by blishko** [comment]
Yes, I will, just need to resolve the conflict in the tests output.

**Comment by blishko** [comment]
Rebased, with updated expected output of command line tests outputting JSON objects.

File: test/libsolidity/smtCheckerTests/types/struct/struct_array_struct_array_storage_unsafe_1.sol [hunk 3]
```diff
@@ -40,8 +40,10 @@ contract C {
 		assert(s1.ts[4].a[5] != 6);
 	}
 }
+// ====
+// SMTIgnoreCex: yes
```
Comment (blishko) [inline test/libsolidity/smtCheckerTests/types/struct/struct_array_struct_array_storage_unsafe_1.sol:44]: Added because of non-determinism in Spacer.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**